### PR TITLE
Final Report version 06

### PR DESCRIPTION
**Pull request by aldojasb**
Hi guys!
Take a look at the final report.
I still need a table to prove why we chose the KNN model. I used a dummy one there.
Let's keep in touch,
See you!

### CODE REVIEW

**Comment by SonQBChau** [comment]
- I see both Rmd and ipynb, am I suppose to complete the python one?
- Can you replace ../results/dummy_table.PNG with the table? I name it best_model.csv
- This report uses the white wine database from “vinho verde” <- do we have a link or reference for this?
- (check the kinks below) <- typo
- Wite wine database <- typo
- In the Analysis, we should add all the package usages. I will give you the list, but you still need to find the BibText

**Comment by SonQBChau** [comment]
Python programming languages and the following Python packages were used to perform the analysis: docopt, pandas, sklearn, altair, numpy, joblib, scipy, pathlib

**Comment by aldojasb** [comment]
-Ops, actually, we don't need the ipynb version. I need to drop off that. Sorry, me bad.
-Yes, I'll change that!
-Thanks for the checks. I'll fix out the typos
-Just send me the packages. In the worst case, we can type it manually.

**Comment by aldojasb** [comment]
-OK! I see the packages now hehe I'll try to google then and find a way to include it haha

**Comment by aldojasb** [comment]
@SonQBChau  about your 'best_model.csv'. Is it an image? In case not, can you create an image of this? That will be easier to include in the final report.

**Comment by SonQBChau** [comment]
No it's not the image, can you load a table through read_csv and kable? The reason is the data in the table can change and I don't want a static image

**Comment by aldojasb** [comment]
I tried before and I got some errors trying to load tables :( but I'll try again, no worries. Let's test it.
Well, I think I need to merge your version to get the files, don't I?

**Comment by SonQBChau** [comment]
- I need someone else to approve my pull request, if nobody does it in the next 30' I will merge it for you.
- About the image, it may not very obvious now, but if I screenshot it, we are unable to auto-generate it in the future. I just try to save us trouble down the line.

**Comment by SonQBChau** [comment]
If you are unable to load the table, I will include the image and we fix it next time

**Comment by SonQBChau** [comment]
You can pull the main repo now, It should have everything you need

**Comment by SonQBChau** [comment]
Please, do not close the pull request by yourself. I need check your work and merge it before you can close it

**Comment by SonQBChau** [comment]
Have you made all the changes and updated the pull request?

**Comment by aldojasb** [comment]
I'm still working on them. I thought I need to close this and then open a new one with my new changes.
When I finish all the changes, what should I do to pull again the files?

**Comment by NikitaShymberg** [comment]
The only reason to close a pull request (that I can think of) is if we decide to not do it at all. If anyone requests any changes, you should simply make them on your forked repo, commit, and push. The pull request will be updated automatically.

File: doc/Quality_white_wine_predictor.md [hunk 1]
```diff
@@ -0,0 +1,245 @@
+Quality white wine predictor
```
Comment (NikitaShymberg) [inline doc/Quality_white_wine_predictor.md:1]: Since we already have the doc/Quality_white_wine_predictor.Rmd file, do we need this one as well?

File: doc/reference.bib [hunk 2]
```diff
@@ -0,0 +1,37 @@
+@Manual{R,
```
Comment (NikitaShymberg) [inline doc/reference.bib:1]: We should also reference all of the python packages that we used, python itself, as well as our dataset.

File: doc/reference.bib [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  }
+  
+@Manual{knitr,
+    title = {knitr: A General-Purpose Package for Dynamic Report Generation in R},
+    author = {Yihui Xie},
+    note = {R package version 1.30.1},
+    url = {https://yihui.org/knitr/},
+  }
+
+@article{milner1969weight,
```
Comment (NikitaShymberg) [inline doc/reference.bib]: I think this is from the individual assignment and should be removed.

File: doc/Quality_white_wine_predictor.md [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+    -   [**Cross Validation**](#cross-validation)
+-   [Results & Discussion](#results--discussion)
+    -   [**Evaluation**](#evaluation)
+    -   [**Discussion**](#discussion)
+    -   [References](#references)
+
+# Summary
+
+This report uses the white wine database from “vinho verde” to predict
+the quality based on physicochemical properties. Quality is a subjective
```
Comment (NikitaShymberg) [inline doc/Quality_white_wine_predictor.md:24]: I love that you split things up into individual lines!

File: doc/Quality_white_wine_predictor.md [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+    -   [**EDA first conclusions**](#eda-first-conclusions)
+    -   [**Cross Validation**](#cross-validation)
+-   [Results & Discussion](#results--discussion)
+    -   [**Evaluation**](#evaluation)
+    -   [**Discussion**](#discussion)
+    -   [References](#references)
+
+# Summary
+
+This report uses the white wine database from “vinho verde” to predict
```
Comment (NikitaShymberg) [inline doc/Quality_white_wine_predictor.md:23]: This is where you can add the citation for the dataset.

File: doc/Quality_white_wine_predictor.md [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+the quality based on physicochemical properties. Quality is a subjective
+measure, given by the average grade of three experts.
+
+Before starting the predictions, the report makes a Explanatory data
```
Comment (NikitaShymberg) [inline doc/Quality_white_wine_predictor.md]: I recommend rephrasing to "the report performs an exploratory data analysis (EDA)".

File: doc/Quality_white_wine_predictor.md [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+
+# Goal
+
+This project aims to determine a model to predict wine quality given
+measurable wine features.
+
+# Introduction
+
+According to experts, wine is differentiated according to its smell,
+flavour, and colour, but most people are not wine experts to say that
```
Comment (NikitaShymberg) [inline doc/Quality_white_wine_predictor.md:49]: Typo: "but most people WHO are not wine experts"

File: doc/Quality_white_wine_predictor.md [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+learning models are important tools to replace human tasks and, in this
+case, a good wine quality prediction can be very useful in the
+certification phase. For example, an automatic predictive system can be
+integrated into a decision support system, helping the speed and quality
+of the performance.
+
+# Data
+
+The wine quality dataset is publicly available on the UCI machine
+learning repository (check the kinks below). The dataset has two files,
```
Comment (NikitaShymberg) [inline doc/Quality_white_wine_predictor.md]: Typo: "check the LINKS below"

File: doc/Quality_white_wine_predictor.md [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+    oxidation of wine
+11. **Residual sugar**: is the amount of sugar remaining after
+    fermentation stops. The key is to have a perfect balance between —
+    sweetness and sourness (wines > 45g/ltrs are sweet)
+
+# Methods
+
+**How to analyze the data** Our task here is to focus on what white wine
+features are important to get the promising result. For the purpose of
+classification model and evaluation of the relevant features, we are
```
Comment (NikitaShymberg) [inline doc/Quality_white_wine_predictor.md]: We are using regression models, not classification models.

Also it might be a good idea to mention here that we are using the `sklearn` library.

File: doc/Quality_white_wine_predictor.md [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+using the following algorithms to perform this task:
+
+1.  **DummyRegressor()**: is one that takes only the value 0 or 1 to
```
Comment (NikitaShymberg) [inline doc/Quality_white_wine_predictor.md]: Maybe I'm misunderstanding, but I don't think that this is how `DummyRegressor()` works.

File: doc/Quality_white_wine_predictor.md [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+    indicate the absence or presence of some categorical effect that may
+    be expected to shift the outcome.
+2.  **Ridge**: is a model tuning method that is used to analyse any data
+    that suffers from multicollinearity. This method performs L2
+    regularization. When the issue of multicollinearity occurs,
+    least-squares are unbiased, and variances are large, this results in
+    predicted values to be far away from the actual values.
+3.  **Random Forest**: is an ensemble learning method for
```
Comment (NikitaShymberg) [inline doc/Quality_white_wine_predictor.md:116]: Did you write these yourself? If not, the source of this information should be mentioned (even if it is just the `sklearn` documentation)

File: doc/Quality_white_wine_predictor.md [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+    classification, regression and other tasks that operates by
+    constructing a multitude of decision trees at training time. For
+    classification tasks, the output of the random forest is the class
+    selected by most trees. For regression tasks, the mean or average
+    prediction of the individual trees is returned
+4.  **KNN**: is an approach to data classification that estimates how
```
Comment (NikitaShymberg) [inline doc/Quality_white_wine_predictor.md:122]: Again, we are doing regression.

File: doc/Quality_white_wine_predictor.md [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+After, data preprocessing is crucial in any data mining process as they
+directly impact success rate of the project. This reduces complexity of
+the data under analysis as data in real world is unclean. we split the
+dataset in two parts, one for training and one for testing. The model
+building and tuning is done in the training set, and then we use the
+test set to predict new values and evaluate the results.
+
+## **EDA first conclusions**
+
+According to our first EDA, we do not have a balanced database, our
```
Comment (NikitaShymberg) [inline doc/Quality_white_wine_predictor.md]: The points made in this paragraph are repeated in the following ones. Maybe it's better to keep this introduction shorter?

File: doc/Quality_white_wine_predictor.md [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+
+## **Cross Validation**
+
+In order to evaluate the best model, we will use the cross validation
+approach to support our decision. The original dataset is partitioned in
+the training set used to train the model, and the test set used to
+predict the values with the trained model. Cross validation partitions
+the training set in the same way and performs the training and
+prediction several times. Then, the result with the best RMSE, accuracy,
+AUC or the chosen metric is selected. This process can be used in
```
Comment (NikitaShymberg) [inline doc/Quality_white_wine_predictor.md]: I would say we just use the r-squared score as our metric here to avoid being too general.

File: doc/Quality_white_wine_predictor.md [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+conjunction to tuning parameters.
+
+# Results & Discussion
+
+## **Evaluation**
+
+After running the models, we used the test-score metrics to evaluate our
+model prediction performance. As we can assess into the table below,
+Model KNN is the best in terms of test-score metrics, with XXX in
```
Comment (NikitaShymberg) [inline doc/Quality_white_wine_predictor.md]: I would talk about the best hyperparameter values here as well.

File: doc/Quality_white_wine_predictor.md [hunk 16]
```diff
@@ -0,0 +1,245 @@
...
+comparison with other models (XXXXX)
+
+![](../results/dummy_table.PNG)<!-- -->
+
+In the context of our business question focusing on the prediction of
+white wine quality, it is reasonable that KNN gives us superior
+“predictions.”
+
+## **Discussion**
```
Comment (NikitaShymberg) [inline doc/Quality_white_wine_predictor.md:234]: I think that our scores weren't very good and that is worth mentioning in here as well.

File: doc/reference.bib (hunk 2 above)
Comment (NikitaShymberg) [reply doc/reference.bib:1]: Also it looks like these references aren't actually used in the .Rmd file.

**Review by NikitaShymberg — COMMENTED** [review]
Thanks Aldo! This is fantastic. I've made quite a few comments, but most of them are minor. My biggest comment is about the references - we need more of them and they should be cited in the document.

**Comment by aldojasb** [comment]
Guys, I have made all the changes above, how can I create a new pull request? I mean, I'm trying to updated my changes here

**Comment by SonQBChau** [comment]
It does update automatically. I'm checking it

**Comment by aldojasb** [comment]
ow, perfect! :)

**Comment by aldojasb** [comment]
I have done two more changes right now.

File: doc/Quality_white_wine_predictor.md (hunk 15 above)
Comment (aldojasb) [reply doc/Quality_white_wine_predictor.md]: ops... I forget that... sorry, my bad :( Give 15min and I'll fix it.

File: doc/reference.bib [hunk 17]
```diff
@@ -7,6 +7,35 @@ @Manual{R
     url = {https://www.R-project.org/},
   }
 
+@article{scikit-learn,
```
Comment (NikitaShymberg) [inline doc/reference.bib:10]: I notice that many things are tagged "@article", I expected this to be @software (though I'm very new to this .bib idea so I might be wrong). What do you think?

Additionally, these things also need to be referenced in the RMD file itself - not just in the .bib file. You can do this by adding `[@label]` tags into the RMD file (e.g. `[@scikit-learn]` in this case).

Comment (aldojasb) [reply doc/reference.bib:10]: about the @software, I'll check that. I'm not sure as well, I just saw some examples in google and copied that haha.

Comment (aldojasb) [reply doc/reference.bib:10]: I've added this in the RMD file. I'm not sure why you cannot see it. take a look at my file:

Comment (aldojasb) [reply doc/reference.bib:10]: image

Comment (aldojasb) [reply doc/reference.bib:10]: I tried to update everything again.

**Comment by SonQBChau** [comment]
maybe you can fix the typo here: DummyRegressor(): <~ DummyRegressor
Otherwise I think it's ok

File: doc/reference.bib (hunk 17 above)
Comment (NikitaShymberg) [reply doc/reference.bib:10]: Oh yes, you're absolutely right, I think I didn't refresh my browser or something.

File: doc/Quality_white_wine_predictor.md (hunk 15 above)
Comment (aldojasb) [reply doc/Quality_white_wine_predictor.md]: done ;-)

**Comment by aldojasb** [comment]
I have included the part about hyperparameters. Try to reload the doc, pelase ;-)

**Comment by aldojasb** [comment]
Guys, I have a hard stop now, but, please, let's keep in touch. I'll come back in 2 hours and check your messages ;-)

**Comment by NikitaShymberg** [comment]
Fantastic, thanks Aldo! I'm happy with this how this looks, so you can merge when you're ready!
